File: src/lib/scala/Slate.Core/src/main/scala/slate/core/apis/ApiBase.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.core.apis

import slate.common._
import slate.common.results.ResultTimed
import slate.core.apis.support.{ApiCallReflect}
import slate.core.common.AppContext

/**
  * Base class for any Api, provides lookup functionality to check for exposed api actions.
  */
class ApiBase {

  protected  val _lookup = new ListMap[String, ApiCallReflect]()

  /**
   * The context of the application. Contains references to the
   * 1. selected environment
   * 2. logger
   * 3. config
   * 4. encryptor
   * 5. authenticator
   * 6. app info and more
   */
  var context:AppContext = null


  /**
   * The api parent container that actually executes the calls on this api
   * and has a register of all the apis.
   */
  var container:ApiContainer = null


  /**
   * hook to allow api to initialize itself
   */
  def init():Unit =
  {
  }


  /**
   * gets a list of all the actions / methods this api supports.
 *
   * @return
   */
  def actions():ListMap[String,ApiCallReflect] =
  {
    _lookup.clone()
  }


  /**
   * whether or not the action exists in this api
    *
    * @param action : e.g. "invite" as in "users.invite"
   * @return
   */
  def contains(action:String):Boolean =
  {
    _lookup.contains(action)
  }


  /**
   * gets the value with the supplied key(action)
    *
    * @param action
   * @return
   */
  def apply(action:String):ApiCallReflect =
  {
    if(!contains(action))
      throw new IllegalArgumentException("action : " + action + " not found")
    _lookup(action)
  }


  /**
   * adds a key/value to this collection
    *
    * @param action
   * @param value
   */
  def update(action:String, value:ApiCallReflect) =
  {
    _lookup(action) = value
  }


  /**
   * benchmarks a function call by tracking the start, end and duration of the call.
    *
    * @param callback : the call to benchmark
   * @return : OperationResultBenchmarked
   */
  def bench(callback:()=>Any ):ResultTimed[Any] =
  {
    var result:Any = null

    val resultTimed = Timer.once("", () => {
      result = callback()
    })

    resultTimed.withData(result)
  }
}

File: src/lib/kotlin/slatekit-apis/src/main/kotlin/slatekit/apis/ApiBase.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.apis

import slatekit.apis.support.FileSupport
import slatekit.common.crypto.Encryptor
import slatekit.common.log.Logger
import slatekit.context.Context


/**
 * Base class for any Api, provides lookup functionality to check for exposed api actions.
 * @param context : The context of the application ( logger, config, encryptor, etc )
 */
abstract class ApiBase(override val context: Context) : FileSupport {
    override val encryptor: Encryptor? = context.enc
    override val logger: Logger? = context.logs.getLogger()
}
